File: Chapter03/Activity3_01/Activity3_01.scala
package Activity3_01

import org.apache.hadoop.io.Text
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import Utilities01.HelperScala.createSession
import Utilities02.HelperScala.{extractRawRecords, parseRawWarc, parseRawWet}
import Utilities02.{WarcRecord, WetRecord}

object Activity3_01 extends App {
  val inputLocWarc = "/Users/a/CC-MAIN-20191013195541-20191013222541-00000.warc"  // ToDo: Modify Path
  val inputLocWet = "/Users/a/CC-MAIN-20191013195541-20191013222541-00000.warc.wet" // ToDo: Modify Path
  implicit val session: SparkSession = createSession(3, "Activity 1")

  val rawRecordsWarc: RDD[Text] = extractRawRecords(inputLocWarc)
  val warcRecords: RDD[WarcRecord] = rawRecordsWarc
    .flatMap(parseRawWarc)
  val rawRecordsWet: RDD[Text] = extractRawRecords(inputLocWet)
  val wetRecords: RDD[WetRecord] = rawRecordsWet
    .flatMap(parseRawWet)
  val pairWarc = warcRecords
    .map(warc => (warc.targetURI, (warc.warcType, warc.dateS, warc.recordID, warc.contentLength, warc.contentType, warc.infoID,
      warc.concurrentTo, warc.ip, warc.payloadDigest, warc.blockDigest, warc.payloadType, warc.htmlContentType, warc.language,
      warc.htmlLength, warc.htmlSource)))
  val pairWet = wetRecords
    .map(wet => (wet.targetURI, wet.plainText))

  val joined = pairWarc
    .join(pairWet)

  println(joined.count())
  Thread.sleep(1000L * 60L * 100L) // For exploring WebUI

}

File: Chapter03/Activity3_01/Activity3_01.py
import time
from Chapter01.utilities01_py.helper_python import create_session
from Chapter02.utilities02_py.helper_python import extract_raw_records, parse_raw_warc, parse_raw_wet

if __name__ == "__main__":
    input_loc_warc = '/Users/a/CC-MAIN-20191013195541-20191013222541-00000.warc'  # ToDo: Modify path
    input_loc_wet = '/Users/a/CC-MAIN-20191013195541-20191013222541-00000.warc.wet'  # ToDo: Modify path
    session = create_session(3, 'Activity 3.1')

    raw_records_warc = extract_raw_records(input_loc_warc, session)
    warc_records = raw_records_warc.flatMap(lambda record: parse_raw_warc(record))
    raw_records_wet = extract_raw_records(input_loc_wet, session)
    wet_records = raw_records_wet.flatMap(lambda record: parse_raw_wet(record))
    pair_warc = warc_records.map(lambda warc: (warc.target_uri, (warc.warc_type, warc.record_id, warc.content_type, warc.block_digest, warc.date_s, warc.content_length, warc.info_id, warc.concurrent_to, warc.ip, warc.payload_digest, warc.payload_type, warc.html_content_type, warc.language, warc.html_length, warc.html_source)))
    pair_wet = wet_records.map(lambda wet: (wet.target_uri, wet.plain_text))

    joined = pair_warc.join(pair_wet, numPartitions=7)

    print(joined.count())
    time.sleep(10 * 60)  # For exploring WebUI
